src/main/java/org/opensearch/neuralsearch/processor/rerank/ByFieldRerankProcessor.java

```diff
@@ -0,0 +1,303 @@
+     * <hr>
+     * <b>This method assumes that the path to the mapping exists as checked by {@link #validateSearchHits(SearchHit[], ActionListener)}</b>
+     * As such no error checking is done in the methods implementing this functionality
+     * @param sourceAsMap the map of maps that contains the <code>targetField</code>
+     */
+    private void removeTargetFieldFromSource(Map<String, Object> sourceAsMap) {
+        String[] keys = targetField.split("\\.");
+        deleteTargetFieldAndEmptyMaps(sourceAsMap, keys);
+    }
+
+    /**
+     * This method performs the deletion of the targetField and emptyMaps in 3 phases
+     * <ol>
+     *     <li>Collect the maps and the respective keys (the key is used to get the inner map) in a stack. It will be used
+     *     to delete empty maps and the target field</li>
+     *     <li>Delete the top most entry, this is guaranteed even when the source mapping is non nested. This is the
+     *     mapping containing the targetField</li>
+     *     <li>Iteratively delete the rest of the maps that have (possibly been) emptied as the result of deleting the targetField</li>
+     * </ol>
+     * <hr>
+     * <b>This method assumes that the path to the mapping exists as checked by {@link #validateSearchHits(SearchHit[], ActionListener)}</b>
+     * As such no error checking is done in the methods implementing this functionality
+     * @param sourceAsMap the map of maps that contains the <code>targetField</code>
+     * @param keys The keys used to traverse the nested map
+     * @implNote You can think of this algorithm as a recursive one the base case is deleting the targetField. The recursive case
+     * is going to the next map along with the respective key. Along the way if it finds a map is empty it will delete it
+     */
+    private void deleteTargetFieldAndEmptyMaps(Map<String, Object> sourceAsMap, String[] keys) {
+        Stack<Tuple<Map<String, Object>, String>> parentMapChildrenKeyTupleStack = new Stack<>();
+
+        Map<String, Object> currentMap = sourceAsMap;
+        String lastKey = keys[keys.length - 1];
+
+        // Collect the parent maps with respective children to use them inside out
+        for (String key : keys) {
+            parentMapChildrenKeyTupleStack.add(new Tuple<>(currentMap, key));
+            if (key.equals(lastKey)) {
+                break;
+            }
+            currentMap = (Map<String, Object>) currentMap.get(key);
+        }
+
+        // Remove the last key this is guaranteed
+        Tuple<Map<String, Object>, String> currentParentMapWithChild = parentMapChildrenKeyTupleStack.pop();
+        Map<String, Object> parentMap = currentParentMapWithChild.v1();
+        String key = currentParentMapWithChild.v2();
+        parentMap.remove(key);
+
+        // Delete the empty maps inside out using the stack to mock a recursive solution
+        while (!parentMapChildrenKeyTupleStack.isEmpty()) {
```

Comment:
I think this loop is not needed. JVM garbage collection will automatically delete this stack.